posts/2023-06-12-streams-in-scala.adoc

```diff
@@ -0,0 +1,214 @@
+
+=== 1st Attempt:
+[source,scala]
+----
+files.map {
+ case (file) =>
+ Source.fromFile(file).getLines().toList.filter(_.contains("Success"))
+  .take(10)
+}
+----
+In the above snippet,we chained getLines and toList which ends up loading the content of the file to memory before filter is called. When we call toList, it evaluates immediately, and only after having the whole data, the filtering is applied. using a strict data structure like List would be a bad idea because of memory usage because the file could be large.
+
+=== 2nd Attempt:
+[source,scala]
+----
+files.map {
+ case (file) => Source.fromFile(file).getLines().to(LazyList).filter(_.contains("Success")).take(10)
+
+}
+----
+In the above snippet, we chained getLines with to(LazyList). With this, the content of the files is not loaded into the memory we then apply the filter and take functions which still don't load the content. The content of the file is only loaded when we call a terminating method like foreach. So the benefit of this is that it would only compute the first ten lines that match the filter predicate so that we don't end up loading everything from file.
+
+[[stream-processing-libraries]]
+== Libraries that implement Streams better
+Some Scala libraries offer enhanced stream processing capabilities compared to the LazyList API. These libraries are implemented following the Reactive stream standard. https://www.reactive-streams.org/[Reactive Streams] is an initiative to provide a standard for asynchronous stream processing with non-blocking back pressure.
+
+=== Reactive Stream
+The Reactive Streams standard establishes two communication channels: an upstream demand channel and a downstream data channel. Publishers follow a request-based approach and only send data when a demand for a certain number of elements arrives through the demand channel. They can then push up to that requested number of elements downstream, either in batches or individually.
+As long as there is outstanding demand, the publisher can continue pushing data to the subscriber as it becomes available. However, when the demand is exhausted, the publisher cannot send data unless prompted by a demand signal from downstream. This mechanism, known as backpressure, ensures controlled flow and prevents overwhelming the subscriber. In response to backpressure, the source can choose to allocate more resources, slow down its production, or even discard data.
+
+image::../media/2023-06-12-streams-in-scala/reactive-stream-representation.png[]
+
+To summarise, handling an un-bounded volume of data in an asynchronous system requires some form of control between the producer and the consumer otherwise we would have overwhelming data sent to the consumer from multiple threads. The Reactive stream standard Introduces a concept of back-pressure which is a means of communication between the producers and the consumer. The reactive stream defines an interface which must be implemented.
```

Comment:
```suggestion
To summarise, handling an un-bounded volume of data in an asynchronous system requires some form of control between the producer and the consumer otherwise we would have overwhelming data sent to the consumer from multiple threads. The Reactive stream standard introduces a concept of back-pressure which is a means of communication between the producers and the consumer. The reactive stream defines an interface which must be implemented.
```